### PR TITLE
implement proper node deletion in tree

### PR DESCRIPTION
**Pull request by gballet**
The PBSS scheme uses reverse-diffs to handle reorgs past the persistent disk layer. For this to work with verkle, `Delete` must be able to completely delete the nodes that were inserted from one tree to the next.

TODO

 * [x] complete leaf deletion to correctly update the commitments
 * [x] fix existing tests
 * [x] add tests to ensure children corner cases are covered
 * [x] swap `delete` and `err`

### CODE REVIEW

File: empty.go [hunk 1]
```diff
@@ -35,8 +35,8 @@ func (Empty) Insert([]byte, []byte, NodeResolverFn) error {
 	return errDirectInsertIntoEmptyNode
 }
 
-func (Empty) Delete([]byte, NodeResolverFn) error {
-	return errors.New("cant delete an empty node")
+func (Empty) Delete([]byte, NodeResolverFn) (error, bool) {
```
Comment (jsign) [inline empty.go:38]: The return order is pretty weird (non-idiomatic), shouldn't be `(bool, error)`?

File: tree.go [hunk 2]
```diff
@@ -513,30 +513,52 @@ func (n *InternalNode) toHashedNode() *HashedNode {
...
 	default:
 		n.cowChild(nChild)
-		return child.Delete(key, resolver)
+		err, delete := child.Delete(key, resolver)
+		if err != nil {
+			return err, false
+		}
+
+		// delete the entire child if instructed to by
+		// the recursive algorightm.
```
Comment (jsign) [inline tree.go:545]: Typo.

File: tree.go [hunk 3]
```diff
@@ -513,30 +513,52 @@ func (n *InternalNode) toHashedNode() *HashedNode {
...
+		if delete {
+			n.children[nChild] = Empty{}
```
Comment (jsign) [inline tree.go:547]: This will be an obvious comment, but we're changing the spec, right?

File: tree.go [hunk 4]
```diff
@@ -1075,15 +1097,76 @@ func (n *LeafNode) updateMultipleLeaves(values [][]byte) {
...
+				isCnempty = false
+				isCempty = false
+				break
+			} else {
+				// i and k[31] were in a different subtree,
+				// so all we can say at this stage, is that
+				// the whole tree isn't empty.
+				// TODO if i < 128, then k[31] >= 128 and
+				// we could skip to 128, but that's an
+				// optimization for later.
```
Comment (jsign) [inline tree.go:1132]: Yes, or be a bit smarter in L1118 (for initialization) to only iterate in the right subtree.
e.g: `i:=k[31]/12*128; i<k[31]/12*128+128; i++` ish
Prob requires some extra `C1 vs C2 != empty` checks to fully complete the optimization.

File: tree.go [hunk 5]
```diff
@@ -1075,15 +1097,76 @@ func (n *LeafNode) updateMultipleLeaves(values [][]byte) {
...
+	// This is done by setting the leaf value
+	// to `nil` at this point, and all the
+	// diff computation will be performed by
+	// updateLeaf since leafToComms supports
+	// nil values.
+	// Note that the value is set to nil by
+	// the method, as it needs the original
+	// value to compute the commitment diffs.
+	n.values[k[31]] = original
+	n.updateLeaf(k[31], nil)
```
Comment (jsign) [inline tree.go:1168]: Doesn't `updateLeaf(...)` updates commitments by using the deletion-marker?
That seems to be a deletion semantics different from all the above logic in this method, which "completely wipes out" the existence of the original value in the commitment.

Looks weird, but I could be missing context. Mentioning it just in case.

**Review by jsign — COMMENTED** [review]
Left some comments.

File: empty.go (hunk 1 above)
Comment (jsign) [reply empty.go:38]: NVM, I see your todo `swap delete and err`.

File: tree.go (hunk 3 above)
Comment (gballet) [reply tree.go:547]: we're not changing the spec in the sense that going forward along the chain, no deletion is possible. But in the case of reorgs, it must be possible to delete values that were inserted in the tree while executing the reorg'ed blocks.

File: tree.go (hunk 4 above)
Comment (gballet) [reply tree.go]: Ah yes good point, I could indeed check that `Cn != zero`, and then only go over one subtree :+1:

File: tree.go (hunk 3 above)
Comment (gballet) [reply tree.go:547]: So to be completely clear here: instead of calling `Delete(key)` in `(*VerkleTrie).Delete`, one should call `Update(key, 0)` to overwrite it with 0s. Geth is responsible for following the spec, this library implements proper deletion as it will be used for the path-based approach to implement deep reorgs.

File: tree.go (hunk 5 above)
Comment (gballet) [reply tree.go]: But the Cn calulcation in the other case was clearly wrong.

File: tree.go (hunk 4 above)
Comment (gballet) [reply tree.go]: not going to fix that today though

File: tree.go (hunk 3 above)
Comment (jsign) [reply tree.go:547]: Ah, OK, so it's an API usage change from the client side compared to before. Makes sense.

File: tree.go [hunk 6]
```diff
@@ -1145,17 +1145,35 @@ func (n *LeafNode) Delete(k []byte, _ NodeResolverFn) (bool, error) {
...
+		// since presumably a lot of values would be deleted at the same
+		// time when reorging. Nonetheless, a reorg is an already complex
+		// operation which is slow no matter what, so ensuring correctness
+		// is more important than
+		var poly [4]Fr
+		toFr(&poly[subtreeindex], cn)
+		n.commitment.Sub(n.commitment, cfg.CommitToPoly(poly[:], 0))
+
+		// Clear the corresponding commitment
+		cn.SetBytes(nil)
```
Comment (jsign) [inline tree.go:1170]: In `tree.go` we usually do `n.c1 != nil` style to detect empty commitments, so this `cn.SetBytes(nil)` is not leaving things in that "style". I I think we might need to set `nil` instead of "clear out" the content. 

So that means falling back to an `if` statement similar to L1153, since we can't do that with `cn` indirection.

**Review by jsign — APPROVED** [review]
